# Intro a Pandas

Pandas es una librería de Python que nos permite trabajar con datos tabulares de manera muy eficiente. 

En el mundo del análisis de datos, Pandas se ha convertido en una herramienta fundamental para científicos de datos, analistas y cualquier persona que trabaje con datos en Python. Con Pandas, puedes realizar operaciones como limpieza, manipulación, visualización y análisis de datos de manera eficiente.

![](https://upload.wikimedia.org/wikipedia/commons/thumb/e/ed/Pandas_logo.svg/1920px-Pandas_logo.svg.png)

## 1. Instalar Pandas

Para instalar Pandas, ejecuta el siguiente comando en tu terminal:

```bash
pip install pandas==<version>
```

In [1]:
import pandas as pd

## 2. Series y Dataframes

Pandas tiene dos estructuras de datos principales: Series y DataFrames.

- **Series**: es un array unidimensional. Es similar a una columna en una tabla de Excel.
- **DataFrame**: es una estructura de datos bidimensional con filas y columnas con etiquetas. Es similar a una tabla de Excel.

### 2.1 Series

Para crear una Serie, puedes pasar una lista de elementos y Pandas creará una Serie con índices numéricos.

In [2]:
serie = pd.Series([1, 2, 3, 4, 5])
serie

0    1
1    2
2    3
3    4
4    5
dtype: int64

### 2.2 DataFrames

Para crear un DataFrame, puedes pasar un diccionario de listas, donde cada clave es el nombre de la columna y cada valor es una lista de elementos.

In [3]:
data = {
    'name': ['María', 'Daniel', 'Marisa', 'David', 'Sabrina'],
    'age': [16, 7, 29, None, 33], 
    'city': ['Madrid', 'Madrid', 'Málaga', 'Málaga', 'Valladolid'],
    'height': [None, 1.20, 1.60, 2.01, 1.83]
}

df = pd.DataFrame(data)
df

,name,age,city,height
0,María,16.0,Madrid,NaN
1,Daniel,7.0,Madrid,1.20
2,Marisa,29.0,Málaga,1.60
3,David,NaN,Málaga,2.01
4,Sabrina,33.0,Valladolid,1.83


Aunque lo más normal es crear DataFrames a partir de archivos CSV, Excel, bases de datos, etc. En este caso, vamos a guardar el DataFrame en un archivo CSV y posteriormente lo leeremos.

In [4]:
CSV_PATH = 'data.csv'

df.to_csv(CSV_PATH, index=False)
df = pd.read_csv(CSV_PATH)
df

,name,age,city,height
0,María,16.0,Madrid,NaN
1,Daniel,7.0,Madrid,1.20
2,Marisa,29.0,Málaga,1.60
3,David,NaN,Málaga,2.01
4,Sabrina,33.0,Valladolid,1.83


## 3. Basics de un Dataframe

Podemos visualizar los primeros/últimos registros de un DataFrame con los métodos `head()` y `tail()`, respectivamente.

In [5]:
df.head(2)

,name,age,city,height
0,María,16.0,Madrid,NaN
1,Daniel,7.0,Madrid,1.2


In [6]:
df.tail(2)

,name,age,city,height
3,David,NaN,Málaga,2.01
4,Sabrina,33.0,Valladolid,1.83


Los dataframes tienen una serie de atributos básicos: `.columns`, `.values`, `.shape`, `.dtypes`, etc. Algunos de ellos te recordarán a los atributos de un array de NumPy.

In [7]:
df.columns

Index(['name', 'age', 'city', 'height'], dtype='str')

In [8]:
df.values

array([['María', 16.0, 'Madrid', nan],
       ['Daniel', 7.0, 'Madrid', 1.2],
       ['Marisa', 29.0, 'Málaga', 1.6],
       ['David', nan, 'Málaga', 2.01],
       ['Sabrina', 33.0, 'Valladolid', 1.83]], dtype=object)

In [9]:
df.shape

(5, 4)

In [10]:
df.dtypes

name          str
age       float64
city          str
height    float64
dtype: object

## 4. Estadísticas básicas

Pandas nos permite obtener la información general de un Dataframe con el método `info()`.

In [11]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 5 entries, 0 to 4
Data columns (total 4 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   name    5 non-null      str    
 1   age     4 non-null      float64
 2   city    5 non-null      str    
 3   height  4 non-null      float64
dtypes: float64(2), str(2)
memory usage: 292.0 bytes




Pandas nos permite obtener estadísticas básicas de un DataFrame con el método `describe()`. Este método nos devuelve un resumen de las estadísticas descriptivas de las **columnas numéricas**.

In [12]:
df.describe()

,age,height
count,4.000000,4.000000
mean,21.250000,1.660000
std,11.954776,0.349571
min,7.000000,1.200000
25%,13.750000,1.500000
50%,22.500000,1.715000
75%,30.000000,1.875000
max,33.000000,2.010000


## 5. Indexación

Recordemos que el indexing consiste en seleccionar un subconjunto de datos de un DataFrame. Recordemos antes cómo era el dataframe original y veamos que Pandas nos permite indexar un DataFrame de diferentes maneras.

In [13]:
df

,name,age,city,height
0,María,16.0,Madrid,NaN
1,Daniel,7.0,Madrid,1.20
2,Marisa,29.0,Málaga,1.60
3,David,NaN,Málaga,2.01
4,Sabrina,33.0,Valladolid,1.83



**Por nombre de columna**: `df['column_name']`

In [14]:
df['name']

0      María
1     Daniel
2     Marisa
3      David
4    Sabrina
Name: name, dtype: str

**Por posición (índice) de la fila**: `df.iloc[0]`


In [15]:
df.iloc[0]

name       María
age         16.0
city      Madrid
height       NaN
Name: 0, dtype: object

**Por etiqueta de la fila**: `df.loc['row_label']`


In [16]:
df.loc[0]  # en este caso, como los indices son los mismos que las posiciones, `loc` es lo mismo que `iloc`

name       María
age         16.0
city      Madrid
height       NaN
Name: 0, dtype: object

**Por etiqueta de fila y columna**: `df.loc['row_label', 'column_name']`

In [17]:
df.loc[0, 'name']

'María'

**Por condición**: `df[df['column_name'] > 0]`


In [18]:
condition = df['age'] > 18
df[condition]  # así se filtran los datos

,name,age,city,height
2,Marisa,29.0,Málaga,1.60
4,Sabrina,33.0,Valladolid,1.83


**Por valores mínimos y máximos**: `df['column_name'].idxmin()` y `df['column_name'].idxmax()`

In [19]:
max_age_row_idx = df["age"].idxmax()
max_age_row = df.loc[max_age_row_idx]
max_age_row

name         Sabrina
age             33.0
city      Valladolid
height          1.83
Name: 4, dtype: object

## 6. Data Cleaning

El data cleaning es un paso fundamental en cualquier análisis de datos. Pandas nos permite realizar operaciones de limpieza de datos de manera muy eficiente. **Es muy común encontrarnos con valores nulos** en nuestros datasets. Pandas nos ofrece métodos para tratar estos valores nulos, como `dropna()`, `fillna()` y `isnull()`. Recordemos primero, el dataset original.

In [20]:
df

,name,age,city,height
0,María,16.0,Madrid,NaN
1,Daniel,7.0,Madrid,1.20
2,Marisa,29.0,Málaga,1.60
3,David,NaN,Málaga,2.01
4,Sabrina,33.0,Valladolid,1.83


Por ejemplo, para ver si hay valores nulos en un DataFrame, podemos usar los metodos `isnull()` ó `isna()`.

In [21]:
df.isnull()

,name,age,city,height
0,False,False,False,True
1,False,False,False,False
2,False,False,False,False
3,False,True,False,False
4,False,False,False,False


Para **eliminar** las filas con valores nulos, podemos usar el método `dropna()`.

In [22]:
df.dropna()

,name,age,city,height
1,Daniel,7.0,Madrid,1.20
2,Marisa,29.0,Málaga,1.60
4,Sabrina,33.0,Valladolid,1.83


Para **rellenar** los valores nulos con un valor específico, podemos usar el método `fillna()`.

In [23]:
df.fillna(0)  # rellena los valores nulos con 0

,name,age,city,height
0,María,16.0,Madrid,0.00
1,Daniel,7.0,Madrid,1.20
2,Marisa,29.0,Málaga,1.60
3,David,0.0,Málaga,2.01
4,Sabrina,33.0,Valladolid,1.83


Estos métodos son aplicables también a las Series (columnas de un DataFrame).

In [24]:
df['age'].fillna(10)

0    16.0
1     7.0
2    29.0
3    10.0
4    33.0
Name: age, dtype: float64

# Ejercicios

Toma el dataframe original y guárdalo como un CSV

Partimos del DataFrame original y lo copiamos a una variable con un nombre descriptivo (`people_df`). Guardamos la ruta en la constante `PEOPLE_CSV_PATH` y exportamos con `to_csv(index=False)` para no escribir el índice como una columna extra.

> ⚠️ El archivo `data.csv` se genera **en local** y está excluido del repositorio mediante `.gitignore`: los datos no deben subirse a GitHub.

In [25]:
people_df = df.copy()
PEOPLE_CSV_PATH = "data.csv"

people_df.to_csv(PEOPLE_CSV_PATH, index=False)

print(f"DataFrame guardado en '{PEOPLE_CSV_PATH}': "
      f"{people_df.shape[0]} filas y {people_df.shape[1]} columnas.")

DataFrame guardado en 'data.csv': 5 filas y 4 columnas.


Ahora lee el archivo CSV

Volvemos a cargar el archivo con `read_csv()`. Al releerlo desde disco, los valores vacíos se interpretan automáticamente como `NaN`.

In [26]:
people_df = pd.read_csv(PEOPLE_CSV_PATH)

print(f"Archivo '{PEOPLE_CSV_PATH}' leído correctamente "
      f"({len(people_df)} registros).")
people_df

Archivo 'data.csv' leído correctamente (5 registros).


,name,age,city,height
0,María,16.0,Madrid,NaN
1,Daniel,7.0,Madrid,1.20
2,Marisa,29.0,Málaga,1.60
3,David,NaN,Málaga,2.01
4,Sabrina,33.0,Valladolid,1.83


Muestra los 3 primeros registros del DataFrame.

`head(n)` devuelve los primeros `n` registros.

In [27]:
first_three_rows = people_df.head(3)

print("Los 3 primeros registros del DataFrame son:")
first_three_rows

Los 3 primeros registros del DataFrame son:


,name,age,city,height
0,María,16.0,Madrid,NaN
1,Daniel,7.0,Madrid,1.2
2,Marisa,29.0,Málaga,1.6


Muestra los 4 últimos registros del DataFrame.

`tail(n)` es el equivalente para el final del DataFrame.

In [28]:
last_four_rows = people_df.tail(4)

print("Los 4 últimos registros del DataFrame son:")
last_four_rows

Los 4 últimos registros del DataFrame son:


,name,age,city,height
1,Daniel,7.0,Madrid,1.20
2,Marisa,29.0,Málaga,1.60
3,David,NaN,Málaga,2.01
4,Sabrina,33.0,Valladolid,1.83


Dime cuantos datos tiene el DataFrame (💡 es lo mismo que calcular el número de filas). ¿en `numpy` vimos 2 formas de calcular este dato, recuerdas?

Igual que en NumPy, tenemos dos formas de saber cuántos registros hay: `len()` sobre el objeto y el primer elemento de la tupla `.shape`.

In [29]:
total_rows_with_len = len(people_df)
total_rows_with_shape = people_df.shape[0]

print(f"El DataFrame tiene {total_rows_with_len} filas calculadas con len() "
      f"y {total_rows_with_shape} filas calculadas con .shape[0]. "
      f"¿Coinciden? {total_rows_with_len == total_rows_with_shape}")

El DataFrame tiene 5 filas calculadas con len() y 5 filas calculadas con .shape[0]. ¿Coinciden? True


Muéstrame las estadísticas básicas del DataFrame.

`describe()` resume las columnas numéricas: conteo, media, desviación típica, mínimo, cuartiles y máximo. Ojo: el conteo (`count`) ignora los nulos, por eso no siempre coincide con el número total de filas.

In [30]:
basic_statistics = people_df.describe()

print("Estadísticas básicas de las columnas numéricas ('age' y 'height'):")
basic_statistics

Estadísticas básicas de las columnas numéricas ('age' y 'height'):


,age,height
count,4.000000,4.000000
mean,21.250000,1.660000
std,11.954776,0.349571
min,7.000000,1.200000
25%,13.750000,1.500000
50%,22.500000,1.715000
75%,30.000000,1.875000
max,33.000000,2.010000


Obtén los datos de la columna 'name' ¿Cual es el tipo de dato de esta columna?

Seleccionamos la columna por su nombre y consultamos su `dtype`. En pandas, las cadenas de texto se almacenan como `object` (o `str` en versiones recientes).

In [31]:
name_column = people_df["name"]
name_column_dtype = name_column.dtype

print(f"La columna 'name' es de tipo '{name_column_dtype}', "
      f"es decir, contiene cadenas de texto.")
name_column

La columna 'name' es de tipo 'str', es decir, contiene cadenas de texto.


0      María
1     Daniel
2     Marisa
3      David
4    Sabrina
Name: name, dtype: str

🌶️ Recuera que una Serie se puede tratar como un array de numpy, por tanto, obtén los datos de la columna "age" y calcula la media. ¿Coincide con el resultado que arroja pandas en la celda superior? (💡 en `numpy` la media se calcula usando `np.mean`)

Una Serie se comporta como un array de NumPy, así que `np.mean()` funciona directamente sobre ella. Ambos métodos **ignoran los valores nulos**, por eso el resultado coincide con el de `describe()`.

In [32]:
import numpy as np

age_column = people_df["age"]

mean_age_with_numpy = np.mean(age_column)
mean_age_with_pandas = age_column.mean()

print(f"Media de edad con numpy:  {mean_age_with_numpy:.2f} años")
print(f"Media de edad con pandas: {mean_age_with_pandas:.2f} años")
print(f"¿Coinciden ambos resultados? "
      f"{np.isclose(mean_age_with_numpy, mean_age_with_pandas)} "
      f"(los dos descartan los valores nulos)")

Media de edad con numpy:  21.25 años
Media de edad con pandas: 21.25 años
¿Coinciden ambos resultados? True (los dos descartan los valores nulos)


Ahora calcula la edad máxima y mínima de la columna "age".

Los métodos `max()` y `min()` de una Serie devuelven el valor extremo, también ignorando los nulos.

In [33]:
max_age = people_df["age"].max()
min_age = people_df["age"].min()

print(f"La edad máxima es de {max_age:.0f} años "
      f"y la edad mínima es de {min_age:.0f} años.")

La edad máxima es de 33 años y la edad mínima es de 7 años.


🌶️ Una vez obtenida la edad mínima, ¿Cual es el nombre de dicha persona?

`idxmin()` no devuelve el valor mínimo, sino **la etiqueta de la fila** donde se encuentra. Con esa etiqueta podemos recuperar cualquier otro dato de la fila usando `loc`.

In [34]:
min_age_row_label = people_df["age"].idxmin()
youngest_person_name = people_df.loc[min_age_row_label, "name"]

print(f"La persona más joven es {youngest_person_name}, "
      f"con {min_age:.0f} años (fila {min_age_row_label}).")

La persona más joven es Daniel, con 7 años (fila 1).


Filtra el DataFrame: muestra las filas donde la edad sea mayor a 25.

Creamos primero una máscara booleana con la condición y después la usamos para filtrar. Separar la condición en su propia variable hace el código más legible y nos permite reutilizarla más adelante.

In [35]:
is_older_than_25 = people_df["age"] > 25
older_than_25_df = people_df[is_older_than_25]

print(f"Personas mayores de 25 años: {len(older_than_25_df)}")
older_than_25_df

Personas mayores de 25 años: 2


,name,age,city,height
2,Marisa,29.0,Málaga,1.60
4,Sabrina,33.0,Valladolid,1.83


Filtra el DataFrame: muestra las filas donde la gente sea de 'Málaga'. (💡 utiliza el operador `==`)

Misma lógica, pero comparando texto con el operador `==`.

In [36]:
is_from_malaga = people_df["city"] == "Málaga"
malaga_df = people_df[is_from_malaga]

print(f"Personas de Málaga: {len(malaga_df)}")
malaga_df

Personas de Málaga: 2


,name,age,city,height
2,Marisa,29.0,Málaga,1.60
3,David,NaN,Málaga,2.01


🌶️ Filtra el Dataframe: filas donde la gente sea de 'Málaga' y mayor de 25. (💡 para sumar dos condiciones se utiliza el operador `and` que se representa en python con `&`, i.e.: 

```python
compose_condition = condition1 & condition2
```

Combinamos las dos máscaras anteriores con `&` (AND elemento a elemento). Importante: los registros con edad nula quedan fuera, porque cualquier comparación con `NaN` devuelve `False`.

In [37]:
is_from_malaga_and_older_than_25 = is_from_malaga & is_older_than_25
malaga_older_than_25_df = people_df[is_from_malaga_and_older_than_25]

print(f"Personas de Málaga mayores de 25 años: {len(malaga_older_than_25_df)}")
malaga_older_than_25_df

Personas de Málaga mayores de 25 años: 1


,name,age,city,height
2,Marisa,29.0,Málaga,1.6


Obtén la máscara de los valores que son nulos en el DataFrame.

`isnull()` devuelve un DataFrame del mismo tamaño con `True` donde hay un nulo. Como `True` vale 1, una doble suma nos da el total de nulos.

In [38]:
null_values_mask = people_df.isnull()
total_null_values = int(null_values_mask.sum().sum())

print(f"El DataFrame contiene {total_null_values} valores nulos en total.")
null_values_mask

El DataFrame contiene 2 valores nulos en total.


,name,age,city,height
0,False,False,False,True
1,False,False,False,False
2,False,False,False,False
3,False,True,False,False
4,False,False,False,False


Ahora obtén la máscara de los valores que NO son nulos en el DataFrame. (💡 busca en la documentación de pandas, o en google, cómo hacerlo)

`notnull()` (o `notna()`) es la máscara inversa: `True` donde **sí** hay dato.

In [39]:
not_null_values_mask = people_df.notnull()
total_not_null_values = int(not_null_values_mask.sum().sum())

print(f"El DataFrame contiene {total_not_null_values} valores no nulos.")
not_null_values_mask

El DataFrame contiene 18 valores no nulos.


,name,age,city,height
0,True,True,True,False
1,True,True,True,True
2,True,True,True,True
3,True,False,True,True
4,True,True,True,True


Elimina los valores nulos del DataFrame.

`dropna()` elimina las filas que tengan **algún** valor nulo. No modifica el DataFrame original: devuelve uno nuevo, por eso lo guardamos en otra variable.

In [40]:
people_without_nulls_df = people_df.dropna()
removed_rows = len(people_df) - len(people_without_nulls_df)

print(f"Se han eliminado {removed_rows} filas con valores nulos: "
      f"pasamos de {len(people_df)} a {len(people_without_nulls_df)} registros.")
people_without_nulls_df

Se han eliminado 2 filas con valores nulos: pasamos de 5 a 3 registros.


,name,age,city,height
1,Daniel,7.0,Madrid,1.20
2,Marisa,29.0,Málaga,1.60
4,Sabrina,33.0,Valladolid,1.83


Rellena todos los valores nulos con 1.

`fillna(valor)` sustituye todos los nulos del DataFrame por el valor indicado.

In [41]:
people_filled_with_ones_df = people_df.fillna(1)

print("Todos los valores nulos han sido sustituidos por 1:")
people_filled_with_ones_df

Todos los valores nulos han sido sustituidos por 1:


,name,age,city,height
0,María,16.0,Madrid,1.00
1,Daniel,7.0,Madrid,1.20
2,Marisa,29.0,Málaga,1.60
3,David,1.0,Málaga,2.01
4,Sabrina,33.0,Valladolid,1.83


Ahora rellena sólo los valores nulos de la columna 'age' con 10.

Si solo queremos tratar una columna, aplicamos `fillna()` sobre esa Serie y reasignamos el resultado. Trabajamos sobre una copia para no alterar el DataFrame original.

In [42]:
people_filled_age_df = people_df.copy()
people_filled_age_df["age"] = people_filled_age_df["age"].fillna(10)

print("Los valores nulos de la columna 'age' se han rellenado con 10:")
people_filled_age_df

Los valores nulos de la columna 'age' se han rellenado con 10:


,name,age,city,height
0,María,16.0,Madrid,NaN
1,Daniel,7.0,Madrid,1.20
2,Marisa,29.0,Málaga,1.60
3,David,10.0,Málaga,2.01
4,Sabrina,33.0,Valladolid,1.83


🌶️ Ahora rellena los valores nulos de la columna 'height' con la media de dicha columna

Rellenar con la media es una estrategia de imputación muy habitual: mantiene la media de la columna intacta y evita perder filas. Calculamos la media (que ignora los nulos) y la usamos como valor de relleno.

In [43]:
mean_height = people_df["height"].mean()

people_filled_height_df = people_df.copy()
people_filled_height_df["height"] = people_filled_height_df["height"].fillna(mean_height)

print(f"La altura media es de {mean_height:.2f} m. "
      f"Los valores nulos de 'height' se han rellenado con ese valor:")
people_filled_height_df

La altura media es de 1.66 m. Los valores nulos de 'height' se han rellenado con ese valor:


,name,age,city,height
0,María,16.0,Madrid,1.66
1,Daniel,7.0,Madrid,1.20
2,Marisa,29.0,Málaga,1.60
3,David,NaN,Málaga,2.01
4,Sabrina,33.0,Valladolid,1.83


## BONUS: Explora los datos del dataset de Titanic

El dataset de Titanic es un dataset muy famoso en el mundo del análisis de datos. Contiene información sobre los pasajeros del Titanic, incluyendo si sobrevivieron o no, su edad, sexo, clase, etc.

In [44]:
TITANIC_CSV_PATH = "https://raw.githubusercontent.com/datasciencedojo/datasets/master/titanic.csv"

titanic_df = pd.read_csv(TITANIC_CSV_PATH)

print(f"Dataset del Titanic cargado desde la URL: "
      f"{titanic_df.shape[0]} pasajeros y {titanic_df.shape[1]} columnas.")
titanic_df.head(5)

Dataset del Titanic cargado desde la URL: 891 pasajeros y 12 columnas.


,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
0,1,0,3,"Braund, Mr. Owen Harris",male,22.0,1,0,A/5 21171,7.2500,NaN,S
1,2,1,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",female,38.0,1,0,PC 17599,71.2833,C85,C
2,3,1,3,"Heikkinen, Miss. Laina",female,26.0,0,0,STON/O2. 3101282,7.9250,NaN,S
3,4,1,1,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",female,35.0,1,0,113803,53.1000,C123,S
4,5,0,3,"Allen, Mr. William Henry",male,35.0,0,0,373450,8.0500,NaN,S


Hazte preguntas como, ¿Qué porcentaje de pasajeros sobrevivió? ¿Cuál es la edad media de los pasajeros? ¿Cuántos pasajeros había en cada clase? ¿Cuál es la edad media de los pasajeros que sobrevivieron y de los que no? ¿Cuál es la edad máxima de los pasajeros que sobrevivieron y de los que no?

> 📌 **Nota sobre el repositorio:** el dataset del Titanic se lee directamente desde su URL con `pd.read_csv()`, de modo que **no se descarga ni se sube ningún CSV al repositorio**. El notebook es reproducible por cualquiera con solo ejecutarlo.

### B.0 Primer vistazo al dataset

Antes de responder a nada, conviene entender qué tenemos entre manos: cuántos registros hay, qué columnas, de qué tipo son y dónde faltan datos.

In [45]:
titanic_df.info()

<class 'pandas.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    str    
 4   Sex          891 non-null    str    
 5   Age          714 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    str    
 9   Fare         891 non-null    float64
 10  Cabin        204 non-null    str    
 11  Embarked     889 non-null    str    
dtypes: float64(2), int64(5), str(5)
memory usage: 83.7 KB


In [46]:
null_values_per_column = titanic_df.isnull().sum()
columns_with_nulls = null_values_per_column[null_values_per_column > 0]

print("Columnas con valores nulos:")
print(columns_with_nulls.to_string())
print(f"\nLa columna 'Age' tiene {null_values_per_column['Age']} nulos sobre "
      f"{len(titanic_df)} pasajeros, algo a tener en cuenta al calcular medias de edad.")

Columnas con valores nulos:
Age         177
Cabin       687
Embarked      2

La columna 'Age' tiene 177 nulos sobre 891 pasajeros, algo a tener en cuenta al calcular medias de edad.


### B.1 ¿Qué porcentaje de pasajeros sobrevivió?

La columna `Survived` está codificada como 0 (no sobrevivió) y 1 (sobrevivió). Al ser una variable binaria, **su media es directamente la tasa de supervivencia**: basta con multiplicarla por 100.

In [47]:
total_passengers = len(titanic_df)
survivors_count = int(titanic_df["Survived"].sum())
survival_rate = titanic_df["Survived"].mean() * 100

print(f"De los {total_passengers} pasajeros del dataset, sobrevivieron {survivors_count}.")
print(f"La tasa de supervivencia es del {survival_rate:.2f}%.")

De los 891 pasajeros del dataset, sobrevivieron 342.
La tasa de supervivencia es del 38.38%.


### B.2 ¿Cuál es la edad media de los pasajeros?

Calculamos la media de la columna `Age`. Recordemos que `mean()` ignora los nulos, así que la media se calcula solo sobre los pasajeros con edad conocida: lo indicamos explícitamente en el resultado para no dar una cifra engañosa.

In [48]:
age_column = titanic_df["Age"]
mean_age = age_column.mean()
passengers_with_known_age = int(age_column.notnull().sum())

print(f"La edad media de los pasajeros es de {mean_age:.2f} años, "
      f"calculada sobre los {passengers_with_known_age} pasajeros con edad conocida "
      f"(de un total de {total_passengers}).")

La edad media de los pasajeros es de 29.70 años, calculada sobre los 714 pasajeros con edad conocida (de un total de 891).


### B.3 ¿Cuántos pasajeros había en cada clase?

`value_counts()` cuenta cuántas veces se repite cada valor de la columna `Pclass`. Ordenamos por clase con `sort_index()` para leerlo de 1ª a 3ª.

In [49]:
passengers_per_class = titanic_df["Pclass"].value_counts().sort_index()

print("Número de pasajeros por clase:")
for passenger_class, passengers_count in passengers_per_class.items():
    class_percentage = passengers_count / total_passengers * 100
    print(f"  - {passenger_class}ª clase: {passengers_count} pasajeros "
          f"({class_percentage:.1f}% del total)")

Número de pasajeros por clase:
  - 1ª clase: 216 pasajeros (24.2% del total)
  - 2ª clase: 184 pasajeros (20.7% del total)
  - 3ª clase: 491 pasajeros (55.1% del total)


### B.4 Edad media de quienes sobrevivieron y de quienes no

Aquí usamos `groupby()`: agrupamos las filas por el valor de `Survived` y aplicamos la media a la columna `Age` dentro de cada grupo.

In [50]:
mean_age_by_survival = titanic_df.groupby("Survived")["Age"].mean()

mean_age_survivors = mean_age_by_survival[1]
mean_age_non_survivors = mean_age_by_survival[0]

print(f"La edad media de los pasajeros que SÍ sobrevivieron es de "
      f"{mean_age_survivors:.2f} años.")
print(f"La edad media de los que NO sobrevivieron es de "
      f"{mean_age_non_survivors:.2f} años.")
print(f"Diferencia: {abs(mean_age_survivors - mean_age_non_survivors):.2f} años.")

La edad media de los pasajeros que SÍ sobrevivieron es de 28.34 años.
La edad media de los que NO sobrevivieron es de 30.63 años.
Diferencia: 2.28 años.


### B.5 Edad máxima de quienes sobrevivieron y de quienes no

Mismo `groupby`, cambiando la función agregada por `max()`.

In [51]:
max_age_by_survival = titanic_df.groupby("Survived")["Age"].max()

max_age_survivors = max_age_by_survival[1]
max_age_non_survivors = max_age_by_survival[0]

print(f"El pasajero de más edad que sobrevivió tenía {max_age_survivors:.0f} años.")
print(f"El pasajero de más edad que no sobrevivió tenía {max_age_non_survivors:.0f} años.")

El pasajero de más edad que sobrevivió tenía 80 años.
El pasajero de más edad que no sobrevivió tenía 74 años.


### B.6 Extra: ¿influyeron el sexo y la clase en la supervivencia?

Aplicando la misma idea del apartado B.1 (la media de una columna binaria es una tasa) sobre grupos, podemos comparar la supervivencia por sexo y por clase.

In [52]:
survival_rate_by_sex = titanic_df.groupby("Sex")["Survived"].mean() * 100

print("Tasa de supervivencia por sexo:")
for sex, rate in survival_rate_by_sex.items():
    print(f"  - {sex}: {rate:.2f}%")

Tasa de supervivencia por sexo:
  - female: 74.20%
  - male: 18.89%


In [53]:
survival_rate_by_class = titanic_df.groupby("Pclass")["Survived"].mean() * 100

print("Tasa de supervivencia por clase:")
for passenger_class, rate in survival_rate_by_class.items():
    print(f"  - {passenger_class}ª clase: {rate:.2f}%")

Tasa de supervivencia por clase:
  - 1ª clase: 62.96%
  - 2ª clase: 47.28%
  - 3ª clase: 24.24%


In [54]:
survival_rate_by_class_and_sex = (
    titanic_df.groupby(["Pclass", "Sex"])["Survived"].mean() * 100
).round(2)

print("Tasa de supervivencia (%) cruzando clase y sexo:")
survival_rate_by_class_and_sex.unstack()

Tasa de supervivencia (%) cruzando clase y sexo:


Sex,female,male
Pclass,,
1,96.81,36.89
2,92.11,15.74
3,50.00,13.54


### Conclusiones del bonus

- Sobrevivió el **38,4% del pasaje** (342 de 891 personas): lo habitual fue no sobrevivir.
- La **edad media ronda los 30 años**, pero falta la edad de casi un 20% de los pasajeros, por lo que conviene tomar el dato con cautela.
- La **3ª clase concentraba a más de la mitad del pasaje** y es, con diferencia, la que peor tasa de supervivencia registra.
- Quienes sobrevivieron eran de media **algo más jóvenes**, aunque la diferencia es de pocos años.
- El factor más determinante no es la edad sino el **sexo**: la tasa de supervivencia de las mujeres multiplica varias veces la de los hombres, y el efecto se acentúa al cruzarlo con la clase (el protocolo *mujeres y niños primero* más el acceso a los botes desde las cubiertas superiores).